### PR TITLE
Adiabatic density profile

### PR DESCRIPTION
**Pull request by gassmoeller**
In order to support simplified density approximations like BA, TALA, ALA we need a reference density profile. This is a first step introducing the profile, which will be used by another PR by @tjhei later on. I expect many test failures, because this PR also solves a small bug in `AdiabaticConditions::InitialProfile`, in which we used a shifted index for the density to compute the next step. The effect is small (10^-4 on temperature and pressure). I will update the PR, when this is ready for review.

### CODE REVIEW

File: source/adiabatic_conditions/function.cc [hunk 1]
```diff
@@ -62,6 +62,25 @@ namespace aspect
...
+      const double z = this->get_geometry_model().depth(p);
+      return function.value(Point<1>(z), 2);
+    }
+
+
+
+    template <int dim>
+    double Function<dim>::density_derivative (const Point<dim> &p) const
+    {
+      // TODO: better eps or make it a user input
```
Comment (tjhei) [inline source/adiabatic_conditions/function.cc]: I am still not sure if we should require the user to specify this derivative. Thoughts?

File: source/adiabatic_conditions/initial_profile.cc [hunk 2]
```diff
@@ -187,23 +201,80 @@ namespace aspect
...
                   ExcInternalError());
           return temperatures.back();
         }
 
-      const unsigned int i = static_cast<unsigned int>(z/delta_z);
-      Assert ((z/delta_z) >= 0, ExcInternalError());
+      if (z < delta_z * std::numeric_limits<double>::epsilon())
+        return temperatures.front();
+
+      const unsigned int i = static_cast<unsigned int>((z/delta_z) * (1. - 2. * std::numeric_limits<double>::epsilon()));
```
Comment (tjhei) [inline source/adiabatic_conditions/initial_profile.cc]: We are now repeating this logic many times in this file. How about moving this into a helper function?

File: source/adiabatic_conditions/function.cc (hunk 1 above)
Comment (tjhei) [inline source/adiabatic_conditions/function.cc]: I am still not sure if we should require the user to specify this derivative. Thoughts?

File: source/adiabatic_conditions/initial_profile.cc [hunk 3]
```diff
@@ -187,23 +201,80 @@ namespace aspect
     {
       const double z = this->get_geometry_model().depth(p);
 
+      Assert (z >= 0, ExcInternalError());
```
Comment (tjhei) [inline source/adiabatic_conditions/initial_profile.cc]: Isn't this too strict? Maybe we should allow being eps below 0. You catch this case below anyways (by returning temperatures.front).

File: source/postprocess/depth_average.cc [hunk 4]
```diff
@@ -300,14 +321,19 @@ namespace aspect
                              "The format in which the output shall be produced. The "
                              "format in which the output is generated also determines "
                              "the extension of the file into which data is written.");
+          const std::string variables =
```
Comment (tjhei) [inline source/postprocess/depth_average.cc]: this is a great idea, I would have done the same. ;-)

File: source/adiabatic_conditions/initial_profile.cc [hunk 5]
```diff
@@ -36,89 +34,99 @@ namespace aspect
...
+          const Tensor <1,dim> g = this->get_gravity_model().gravity_vector(representative_point);
 
-          MaterialModel::MaterialModelInputs<dim> in(1, n_compositional_fields);
-          MaterialModel::MaterialModelOutputs<dim> out(1, n_compositional_fields);
           in.position[0] = representative_point;
-          in.temperature[0] = temperatures[i-1];
-          in.pressure[0] = pressures[i-1];
+          in.temperature[0] = temperatures[i];
+          in.pressure[0] = pressures[i];
           in.velocity[0] = Tensor <1,dim> ();
```
Comment (tjhei) [inline source/adiabatic_conditions/initial_profile.cc]: can we move this and the strainrate, etc. outside the loop?

File: source/adiabatic_conditions/initial_profile.cc [hunk 6]
```diff
@@ -144,65 +152,97 @@ namespace aspect
...
-      Assert (i+1 < temperatures.size(), ExcInternalError());
+      if (z <= 0)
+        {
+          Assert (z >= -delta_z, ExcInternalError());
+          return property.front();
+        }
 
+      const double floating_index = z/delta_z;
+      const unsigned int i = static_cast<unsigned int>(floating_index);
+      if (std::abs(floating_index-std::round(floating_index)) < 1e-6)
```
Comment (tjhei) [inline source/adiabatic_conditions/initial_profile.cc]: can you add a comment here please?

File: source/adiabatic_conditions/initial_profile.cc [hunk 7]
```diff
@@ -144,65 +152,97 @@ namespace aspect
...
+        return property[i];
+
+      Assert (i+1 < temperatures.size(), ExcInternalError());
```
Comment (tjhei) [inline source/adiabatic_conditions/initial_profile.cc]: ``property.size``?

File: source/material_model/latent_heat.cc [hunk 8]
```diff
@@ -42,7 +42,7 @@ namespace aspect
       // note: the pressure case for adiabatic conditions initialized and
       // the pressure case for non-initialized adiabatic conditions
       // are the same, so the pressure case is included in the if statement for adiabatic conditions.
-      if (this->get_adiabatic_conditions().is_initialized() || use_depth ==false)
+      if (use_depth ==false)
```
Comment (tjhei) [inline source/material_model/latent_heat.cc]: Can you fix the spacing around ``==`` or use ``!use_depth``?

**Comment by gassmoeller** [comment]
I addressed the comments. For now I would leave the TODO about the better epsilon for the function derivative to make progress with the other parts.

**Pull request merged**